File: frontend/src/components/benchmark/content/adamic_adar/implementations/adamic_adar.js
class AdamicAdar {
    constructor() {
        this.adjacencyList = new Map();
    }

    addEdge(u, v) {
        if (!this.adjacencyList.has(u)) {
            this.adjacencyList.set(u, new Set());
        }
        if (!this.adjacencyList.has(v)) {
            this.adjacencyList.set(v, new Set());
        }
        this.adjacencyList.get(u).add(v);
        this.adjacencyList.get(v).add(u);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    adamicAdarScore(u, v) {
        if (!this.adjacencyList.has(u) || !this.adjacencyList.has(v)) {
            return 0.0;
        }

        let score = 0.0;
        const neighborsU = this.adjacencyList.get(u);
        const neighborsV = this.adjacencyList.get(v);

        for (const node of neighborsU) {
            if (neighborsV.has(node)) {
                const degree = this.adjacencyList.get(node).size;
                if (degree > 1) {
                    score += 1.0 / Math.log(degree);
                }
            }
        }
        return score;
    }

    scoreAllPairs() {
        const scores = [];
        const nodes = Array.from(this.adjacencyList.keys());

        for (let i = 0; i < nodes.length; i++) {
            for (let j = i + 1; j < nodes.length; j++) {
                const u = nodes[i];
                const v = nodes[j];
                const score = this.adamicAdarScore(u, v);
                if (score > 0) {
                    scores.push({ pair: [u, v], score: score });
                }
            }
        }

        scores.sort((a, b) => b.score - a.score);
        return scores;
    }

    topKPredictions(k = 10) {
        const scores = this.scoreAllPairs();
        return scores.slice(0, k);
    }
}

module.exports = AdamicAdar;

File: frontend/src/components/benchmark/content/adamic_adar/implementations/adamic_adar.rs
use std::collections::{HashMap, HashSet};

pub struct AdamicAdar {
    adjacency_list: HashMap<i32, HashSet<i32>>,
}

impl AdamicAdar {
    pub fn new() -> Self {
        AdamicAdar {
            adjacency_list: HashMap::new(),
        }
    }

    pub fn add_edge(&mut self, u: i32, v: i32) {
        self.adjacency_list.entry(u).or_insert_with(HashSet::new).insert(v);
        self.adjacency_list.entry(v).or_insert_with(HashSet::new).insert(u);
    }

    pub fn build_from_edges(&mut self, edges: &[(i32, i32)]) {
        for &(u, v) in edges {
            self.add_edge(u, v);
        }
    }

    pub fn adamic_adar_score(&self, u: i32, v: i32) -> f64 {
        if let (Some(neighbors_u), Some(neighbors_v)) = (self.adjacency_list.get(&u), self.adjacency_list.get(&v)) {
            let mut score = 0.0;
            for &node in neighbors_u {
                if neighbors_v.contains(&node) {
                    if let Some(neighbors) = self.adjacency_list.get(&node) {
                        let degree = neighbors.len();
                        if degree > 1 {
                            score += 1.0 / (degree as f64).ln();
                        }
                    }
                }
            }
            return score;
        }
        0.0
    }

    pub fn score_all_pairs(&self) -> Vec<((i32, i32), f64)> {
        let mut scores = Vec::new();
        let nodes: Vec<i32> = self.adjacency_list.keys().cloned().collect();

        for i in 0..nodes.len() {
            for j in (i + 1)..nodes.len() {
                let u = nodes[i];
                let v = nodes[j];
                let score = self.adamic_adar_score(u, v);
                if score > 0.0 {
                    scores.push(((u, v), score));
                }
            }
        }

        scores.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
        scores
    }

    pub fn top_k_predictions(&self, k: usize) -> Vec<((i32, i32), f64)> {
        let scores = self.score_all_pairs();
        scores.into_iter().take(k).collect()
    }
}